File: all-queens.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void print_board(int n, int *board) {
    if (board == NULL)
        printf("%s", (char *) board);
    else 
        for (int i=0; i<n; i++) {
            for (int j=0; j<n; j++) {
                if (board[i] == j)
                    printf("x");
                else
                    printf("o");
            }
            printf("\n");
        }
    printf("\n");
}

int is_attacked(int n, int *board, int i0, int j0) {
    for (int i=0; i<n; i++) {
        if (board[i] == -1)
            continue;
        if (board[i] == j0)
            return 1;
        
        if ((i - board[i] == i0 - j0) || (i + board[i] == i0 + j0))
            return 1;
    }
    return 0;
}

void rec(int n, int *board, int row, int *total_sols, int verbose) {
    if (row >= n) {
        (*total_sols)++;
        if (verbose)
            print_board(n, board);
    }

    for (int col=0; col<n; col++) {
        if (!is_attacked(n, board, row, col)) {
            board[row] = col;
            rec(n, board, row+1, total_sols, verbose);
            board[row] = -1;
        }
    }
}

int all_n_queens(int n, int verbose) {
    int *board;
    int total_sols = 0;

    board = malloc(sizeof(int)*n);
    memset(board, -1, sizeof(int)*n);
    rec(n, board, 0, &total_sols, verbose);
    return total_sols;
}

int main(int argc, char **argv) {
    int *sol;
    int n;
    int print;
    int pos_n;

    if (argc < 2) {
        printf("Usage: program [-v] n\n");
        return 1;
    }

    print = 0;
    if (strcmp(argv[1], "-v") == 0) {
        print = 1;
        pos_n = 2;
    } else
        pos_n = 1;

    if (argc != pos_n + 1) {
        printf("Usage: program [-v] n\n");
        return 1;
    }

    n = atoi(argv[pos_n]);
    int i = all_n_queens(n, print);

    printf("%d\n", i);
    return 0;
}

File: all-queens.hs
import System.Environment
import Prelude

is_attacked :: [Int] -> Int -> Int -> Bool
is_attacked board i0 j0 = aux_rec board i0 j0 0 where
    aux_rec board i0 j0 i
        | i == length board = False
        | otherwise           = ((b_i>=0) &&
                                    ((b_i==j0)||(i-b_i == i0-j0)||(i+b_i == i0+j0)))
                                || aux_rec board i0 j0 (i+1) where b_i = board !! i

put_col :: Int -> Int -> [Int] -> [Int]
put_col row col (x:xs) = if row==0 then col:xs else x:(put_col (row-1) col xs)

all_n_queens :: Int -> Int
all_n_queens n = rec_aux board n 0 where
    board = take (n) (repeat (-1))
    rec_aux :: [Int] -> Int -> Int -> Int
    rec_aux board n row
        | row >= n  = 1
        | otherwise = foldl (\x col ->
            x+(if (is_attacked board row col) then 0 else (rec_aux (put_col row col board) n (row+1)))
                ) 0 [0..n-1]

board = take (4) (repeat (-1))
main :: IO() = do
    args <- getArgs
    if length args /= 1 then (print "Usage: program [-v] n") else (
                    let n = (args !! 0)  in (print $ all_n_queens (read n)))
                        --(put_col 1 1 board)))